<div id='2.2' style='background-color:green;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:white;font-family:Futura;font-size: 30px'> RFM Score and Customer Behavior </h3>
</div>

### The combination of Recency, Frequency and Monetary forms the RFM score, a numerical representation of a customer's overall engagement and value to the business. The RFM score is often used in customer segmentation and targeted marketing strategies.

* Recency (R): Refers to how recently a customer has made a purchase or interacted with the business. It's often measured as the time since the customer's last transaction.

* Frequency (F): Represents how often a customer makes a purchase or interacts with the business. It's typically measured as the total number of transactions or interactions over a specific period.

* Monetary (M): Indicates the monetary value of a customer's transactions or spending with the business. It's measured as the total amount of money spent by the customer over a specific period.

In [2]:
# Importing libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

In [3]:
df = pd.read_csv('sales.csv')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 375187 entries, 0 to 375186
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    375187 non-null  int64  
 1   StockCode    375187 non-null  object 
 2   Description  375187 non-null  object 
 3   Quantity     375187 non-null  int64  
 4   InvoiceDate  375187 non-null  object 
 5   UnitPrice    375187 non-null  float64
 6   CustomerID   375187 non-null  float64
 7   Country      375187 non-null  object 
dtypes: float64(2), int64(2), object(4)
memory usage: 22.9+ MB


In [4]:
df.describe()

InvoiceNo       Quantity      UnitPrice     CustomerID
count  375187.000000  375187.000000  375187.000000  375187.000000
mean   560720.262192      13.584978       2.318388   15306.779446
std     13104.443738     184.647865       1.825628    1710.647733
min    536365.000000       1.000000       0.001000   12346.000000
25%    549291.000000       2.000000       1.060000   13975.000000
50%    562031.000000       6.000000       1.650000   15187.000000
75%    572194.000000      12.000000       2.950000   16808.000000
max    581587.000000   80995.000000       8.290000   18287.000000

In [5]:
df.nunique()

InvoiceNo      18224
StockCode       3479
Description     3674
Quantity         301
InvoiceDate    17017
UnitPrice        275
CustomerID      4311
Country           37
dtype: int64

In [4]:
# convert the datatype of invoicedate into datetime

df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [11]:
df.dtypes

InvoiceNo               int64
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [12]:
# checking for null values

df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [7]:
# checking for duplicates

df[df.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
482        536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
492        536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
502        536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
504        536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
520        536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...           ...       ...                                ...       ...   
374972     581538     22068        BLACK PIRATE TREASURE CHEST         1   
374986     581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
374989     581538     22992             REVOLVER WOODEN RULER          1   
374996     581538     22694                       WICKER STAR          1   
374998     581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

               InvoiceDate  UnitPrice  CustomerID         Country  
482    2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
492    2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
502    2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
504    2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
520    2010-12-01 11:49:00       2.95     17920.0  United Kingdom  
...                    ...        ...         ...             ...  
374972 2011-12-09 11:34:00       0.39     14446.0  United Kingdom  
374986 2011-12-09 11:34:00       2.49     14446.0  United Kingdom  
374989 2011-12-09 11:34:00       1.95     14446.0  United Kingdom  
374996 2011-12-09 11:34:00       2.10     14446.0  United Kingdom  
374998 2011-12-09 11:34:00       2.08     14446.0  United Kingdom  

[5021 rows x 8 columns]

We can identify that this dataset contains 5021 rows and to avoid data leakage we must delete these duplicated rows.

In [8]:
df_new = df.drop_duplicates()
df_new

InvoiceNo StockCode                          Description  Quantity  \
0          536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1          536365     71053                  WHITE METAL LANTERN         6   
2          536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3          536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4          536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...           ...       ...                                  ...       ...   
375182     581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
375183     581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
375184     581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
375185     581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
375186     581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
375182 2011-12-09 12:50:00       0.85     12680.0          France  
375183 2011-12-09 12:50:00       2.10     12680.0          France  
375184 2011-12-09 12:50:00       4.15     12680.0          France  
375185 2011-12-09 12:50:00       4.15     12680.0          France  
375186 2011-12-09 12:50:00       4.95     12680.0          France  

[370166 rows x 8 columns]

In [9]:
# most recent date

df_new['InvoiceDate'].max()

Timestamp('2011-12-09 12:50:00')

In [10]:
# taking december 11 as current date, so we can find recency

current_date = "2011-12-11"
current_date = pd.to_datetime(current_date)
print(type(current_date))

<class 'pandas._libs.tslibs.timestamps.Timestamp'>


In [11]:
# adding an extra column named Total_price

df_new['Total_price'] = df_new['Quantity'] * df_new['UnitPrice']
df_new.head()

C:\Users\lenovo\AppData\Local\Temp\ipykernel_19712\3137562184.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['Total_price'] = df_new['Quantity'] * df_new['UnitPrice']


InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Total_price  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom        15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom        22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom        20.34

<div id='2.2' style='background-color:lightpink;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> RFM (Recency,Frequency,monetary)</h3>
</div>

<div id='2.2' style='background-color:lightblue;text-align:center;padding: 5px; line-height:0.5;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 20px'> RECENCY </h3>
</div>

RECENCY refers to the measure of how recently a customer has made a purchase or engaged with a product or service. 

In [30]:
recency_df = df_new.groupby('CustomerID')['InvoiceDate'].max().reset_index()
recency_df['Recency'] = (current_date - recency_df['InvoiceDate']).dt.days
recency_df = recency_df[['CustomerID','Recency']]
recency_df = recency_df.set_index('CustomerID')
recency_df

Recency
CustomerID         
12346.0         326
12347.0           3
12348.0          76
12349.0          19
12350.0         311
...             ...
18280.0         278
18281.0         181
18282.0           8
18283.0           4
18287.0          43

[4311 rows x 1 columns]

<div id='2.2' style='background-color:lightblue;text-align:center;padding: 5px; line-height:0.5;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 20px'> FREQUENCY </h3>
</div>

Frequency refers to the measure of how often a customer engages with a product, service, or makes a purchase within a specific period.

In [29]:
frequency_df = df_new.groupby('CustomerID')['InvoiceNo'].count().reset_index()
frequency_df = frequency_df.set_index('CustomerID')
frequency_df

InvoiceNo
CustomerID           
12346.0             1
12347.0           176
12348.0            27
12349.0            66
12350.0            16
...               ...
18280.0             9
18281.0             5
18282.0             9
18283.0           717
18287.0            69

[4311 rows x 1 columns]

<div id='2.2' style='background-color:lightblue;text-align:center;padding: 5px; line-height:0.5;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 20px'> MONETARY </h3>
</div> 

Monetary refers to the measure of how much money a customer has spent on a product or service over a specific period.

In [28]:
monetary_df = df_new.groupby('CustomerID')['Total_price'].sum().reset_index()
monetary_df = monetary_df.set_index('CustomerID')
monetary_df

Total_price
CustomerID             
12346.0        77183.60
12347.0         4078.95
12348.0         1437.24
12349.0         1287.15
12350.0          294.40
...                 ...
18280.0          160.70
18281.0           46.92
18282.0          127.05
18283.0         2002.63
18287.0         1803.28

[4311 rows x 1 columns]

<div id='2.2' style='background-color:lightblue;text-align:center;padding: 5px; line-height:0.5;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 20px'> RFM  </h3>
</div>

In [32]:
RFM_df = pd.concat([recency_df,frequency_df,monetary_df],axis=1)
recency_df.columns=["Recency"]
frequency_df.columns=["Frequency"]
monetary_df.columns=["Monetary"]
RFM_df

Recency  Frequency  Monetary
CustomerID                              
12346.0         326          1  77183.60
12347.0           3        176   4078.95
12348.0          76         27   1437.24
12349.0          19         66   1287.15
12350.0         311         16    294.40
...             ...        ...       ...
18280.0         278          9    160.70
18281.0         181          5     46.92
18282.0           8          9    127.05
18283.0           4        717   2002.63
18287.0          43         69   1803.28

[4311 rows x 3 columns]

#### Using RFM analysis, customers with the lowest recency, highest frequency and highest monetary amounts are considered as top customers

In [34]:
r_labels, f_labels, m_labels = range(5,0,-1),range(1,6),range(1,6)

RFM_df['r_score'] = pd.qcut(RFM_df['Recency'], q=5, labels=r_labels).astype(int)
RFM_df['f_score'] = pd.qcut(RFM_df['Frequency'], q=5, labels=f_labels).astype(int)
RFM_df['m_score'] = pd.qcut(RFM_df['Monetary'], q=5, labels=m_labels).astype(int)

RFM_df['rfm_sum'] = RFM_df['r_score'] + RFM_df['f_score'] + RFM_df['m_score']
RFM_df

Recency  Frequency  Monetary  r_score  f_score  m_score  rfm_sum
CustomerID                                                                  
12346.0         326          1  77183.60        1        1        5        7
12347.0           3        176   4078.95        5        5        5       15
12348.0          76         27   1437.24        2        2        4        8
12349.0          19         66   1287.15        4        4        4       12
12350.0         311         16    294.40        1        2        2        5
...             ...        ...       ...      ...      ...      ...      ...
18280.0         278          9    160.70        1        1        1        3
18281.0         181          5     46.92        1        1        1        3
18282.0           8          9    127.05        5        1        1        7
18283.0           4        717   2002.63        5        5        5       15
18287.0          43         69   1803.28        3        4        4       11

[4311 rows x 7 columns]

<div id='2.2' style='background-color:lightpink;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> RFM Customer Segments</h3>
</div>

* Champions - bought recently, buy often and spend the most
* Loyal Customers - spend good money and often, responsive to promotions
* Potential Loyalist - recent customers, but spent a good amount and bought more than once
* New Customers - bought most recently, but not often
* Promising - recent shoppers, but haven’t spent much
* Needing Attention - above average recency, frequency and monetary values; may not have bought very recently though
* About To Sleep - below average recency, frequency and monetary values; will lose them if not reactivated
* At Risk - spent big money and purchased often but long time ago; need to bring them back
* Can't Loose Them - made biggest purchases, and often but haven’t returned for a long time
* Hibernating - last purchase was long back, low spenders and low number of orders

In [35]:
segt_map = {
    r'[1-2][1-2]': 'hibernating',
    r'[1-2][3-4]': 'at risk',
    r'[1-2]5': 'can\'t loose',
    r'3[1-2]': 'about to sleep',
    r'33': 'need attention',
    r'[3-4][4-5]': 'loyal customers',
    r'41': 'promising',
    r'51': 'new customers',
    r'[4-5][2-3]': 'potential loyalists',
    r'5[4-5]': 'champions'
}

RFM_df['Segment'] = RFM_df['r_score'].map(str) + RFM_df['f_score'].map(str)
RFM_df['Segment'] = RFM_df['Segment'].replace(segt_map, regex=True)
RFM_df

Recency  Frequency  Monetary  r_score  f_score  m_score  rfm_sum  \
CustomerID                                                                     
12346.0         326          1  77183.60        1        1        5        7   
12347.0           3        176   4078.95        5        5        5       15   
12348.0          76         27   1437.24        2        2        4        8   
12349.0          19         66   1287.15        4        4        4       12   
12350.0         311         16    294.40        1        2        2        5   
...             ...        ...       ...      ...      ...      ...      ...   
18280.0         278          9    160.70        1        1        1        3   
18281.0         181          5     46.92        1        1        1        3   
18282.0           8          9    127.05        5        1        1        7   
18283.0           4        717   2002.63        5        5        5       15   
18287.0          43         69   1803.28        3        4        4       11   

                    Segment  
CustomerID                   
12346.0         hibernating  
12347.0           champions  
12348.0         hibernating  
12349.0     loyal customers  
12350.0         hibernating  
...                     ...  
18280.0         hibernating  
18281.0         hibernating  
18282.0       new customers  
18283.0           champions  
18287.0     loyal customers  

[4311 rows x 8 columns]

In [36]:
# Add country column to the RFM dataframe
df_rfm = pd.merge(RFM_df, df_new[['CustomerID', 'Country']], on='CustomerID', how='left')
df_rfm = df_rfm.drop_duplicates()
df_rfm = df_rfm.dropna(subset=['Country'])
df_rfm = df_rfm.reset_index(drop=True) #Rest the index value
df_rfm

CustomerID  Recency  Frequency  Monetary  r_score  f_score  m_score  \
0        12346.0      326          1  77183.60        1        1        5   
1        12347.0        3        176   4078.95        5        5        5   
2        12348.0       76         27   1437.24        2        2        4   
3        12349.0       19         66   1287.15        4        4        4   
4        12350.0      311         16    294.40        1        2        2   
...          ...      ...        ...       ...      ...      ...      ...   
4314     18280.0      278          9    160.70        1        1        1   
4315     18281.0      181          5     46.92        1        1        1   
4316     18282.0        8          9    127.05        5        1        1   
4317     18283.0        4        717   2002.63        5        5        5   
4318     18287.0       43         69   1803.28        3        4        4   

      rfm_sum          Segment         Country  
0           7      hibernating  United Kingdom  
1          15        champions         Iceland  
2           8      hibernating         Finland  
3          12  loyal customers           Italy  
4           5      hibernating          Norway  
...       ...              ...             ...  
4314        3      hibernating  United Kingdom  
4315        3      hibernating  United Kingdom  
4316        7    new customers  United Kingdom  
4317       15        champions  United Kingdom  
4318       11  loyal customers  United Kingdom  

[4319 rows x 10 columns]

<div id='2.2' style='background-color:lightgreen;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> CHAMPIONS</h3>
</div>

In [37]:
champions_df = df_rfm[df_rfm['Segment']=='champions']
champions_df

CustomerID  Recency  Frequency  Monetary  r_score  f_score  m_score  \
1        12347.0        3        176   4078.95        5        5        5   
15       12362.0        4        240   4377.28        5        5        5   
17       12364.0        8         80   1186.20        5        4        4   
31       12381.0        5         75   1276.86        5        4        4   
58       12417.0        4        150   2417.10        5        5        5   
...          ...      ...        ...       ...      ...      ...      ...   
4286     18237.0        3         56    911.85        5        4        4   
4289     18241.0       10         99   1916.24        5        4        5   
4291     18245.0        8        167   2404.51        5        5        5   
4308     18272.0        3        166   3078.58        5        5        5   
4317     18283.0        4        717   2002.63        5        5        5   

      rfm_sum    Segment         Country  
1          15  champions         Iceland  
15         15  champions         Belgium  
17         13  champions         Belgium  
31         13  champions          Norway  
58         15  champions         Belgium  
...       ...        ...             ...  
4286       13  champions  United Kingdom  
4289       14  champions  United Kingdom  
4291       15  champions  United Kingdom  
4308       15  champions  United Kingdom  
4317       15  champions  United Kingdom  

[623 rows x 10 columns]

### These 623 customers are highly engaged and have made recent and frequent purchases.

* Consider implementing a loyalty program or providing exclusive offers to further encourage their engagement.
* Personalize communication to this segment based on their past behaviors and preferences. 
* Identify related or complementary products to the ones they have purchased.
* Implement targeted cross-selling and upselling strategies to increase the average transaction value from these customers.
* Provide exclusive offers, early access to sales, or limited-time promotions to further incentivize their purchasing behavior and make them feel valued.
* Seek feedback from these champion customers.
* Encourage these loyal customers to refer friends and family through referral programs.

<div id='2.2' style='background-color:lightgreen;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> NEW CUSTOMERS </h3>
</div>

In [38]:
newCustomers_df = df_rfm[df_rfm['Segment']=='new customers']
newCustomers_df

CustomerID  Recency  Frequency  Monetary  r_score  f_score  m_score  \
19       12367.0        5         10    150.90        5        1        1   
84       12442.0        4         11    144.06        5        1        1   
176      12558.0        8         11    269.96        5        1        2   
201      12587.0        8          3    104.00        5        1        1   
264      12660.0       12         12    227.10        5        1        2   
...          ...      ...        ...       ...      ...      ...      ...   
4133     18030.0        5          6    123.95        5        1        1   
4151     18058.0       10          3    170.16        5        1        1   
4237     18174.0        8          1    104.00        5        1        1   
4309     18273.0        3          3    204.00        5        1        1   
4316     18282.0        8          9    127.05        5        1        1   

      rfm_sum        Segment         Country  
19          7  new customers         Denmark  
84          7  new customers           Spain  
176         8  new customers             USA  
201         7  new customers         Finland  
264         8  new customers          France  
...       ...            ...             ...  
4133        7  new customers  United Kingdom  
4151        7  new customers  United Kingdom  
4237        7  new customers  United Kingdom  
4309        7  new customers  United Kingdom  
4316        7  new customers  United Kingdom  

[64 rows x 10 columns]

### Dealing with new customers who bought most recently but not often requires a specific approach to encourage their engagement and increase their frequency of purchases.

* Provide special welcome offers or discounts to incentivize new customers to make additional purchases. This could be a percentage off their next purchase, free shipping, or a buy-one-get-one offer.
* Share educational content about your products or services to help new customers understand their value. 
* Set up a series of engagement emails that are triggered after the first purchase. These emails can include product recommendations, tips on using the purchased items, and exclusive offers to encourage repeat purchases.
* Send surveys to new customers to gather feedback on their experience and understand any barriers to repeat purchases. Use this information to make improvements and tailor future marketing efforts.
* Leverage data analytics to provide personalized product recommendations based on the customer's first purchase. Showcasing relevant products increases the likelihood of them making additional purchases.
* Use social media platforms to engage with new customers.

<div id='2.2' style='background-color:lightgreen;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> AT RISK </h3>
</div>

In [39]:
risk_df = df_rfm[df_rfm['Segment']=='at risk']
risk_df

CustomerID  Recency  Frequency  Monetary  r_score  f_score  m_score  \
7        12354.0      233         49    925.95        1        3        4   
9        12356.0      246         50   1911.40        1        3        5   
27       12377.0      316         75   1488.42        1        4        4   
29       12379.0       82         38    777.24        2        3        3   
32       12383.0      185         83   1382.96        1        4        4   
...          ...      ...        ...       ...      ...      ...      ...   
4283     18232.0       82         34    549.47        2        3        3   
4287     18239.0      219         85   1473.71        1        4        4   
4290     18242.0       73         74   2133.64        2        4        5   
4293     18248.0      123         44    684.32        2        3        3   
4305     18265.0       73         38    658.01        2        3        3   

      rfm_sum  Segment         Country  
7           8  at risk           Spain  
9           9  at risk        Portugal  
27          9  at risk     Switzerland  
29          8  at risk         Belgium  
32          9  at risk         Belgium  
...       ...      ...             ...  
4283        8  at risk  United Kingdom  
4287        9  at risk  United Kingdom  
4290       11  at risk  United Kingdom  
4293        8  at risk  United Kingdom  
4305        8  at risk  United Kingdom  

[555 rows x 10 columns]

### There is 555 customers are at risk and require targeted efforts to bring them back.

* Design targeted email or marketing campaigns specifically for "At Risk" customers. Highlight special offers, new products, or exclusive discounts to entice them to make a purchase.
* Create a sense of urgency with limited-time offers or promotions. Communicate that these offers are exclusive and time-sensitive, encouraging quick action.
* Send surveys to gather feedback on why they haven't made recent purchases. Understanding their concerns or reasons for disengagement can help you address specific issues and improve their experience.
* Consider placing "At Risk" customers in an exclusive VIP program with special benefits and perks. This could include early access to sales, exclusive events, or additional discounts.
* Leverage data to provide personalized product recommendations based on their past purchases. Showcasing items they might be interested in can reignite their interest.

<div id='2.2' style='background-color:lightgreen;text-align:center;padding: 15px; line-height:1;'>
    <h3 style='color:black;font-family:"Courier New", Courier, monospace;font-size: 25px'> HIBERNATING </h3>
</div>

In [40]:
hibernating_df = df_rfm[df_rfm['Segment']=='hibernating']
hibernating_df

CustomerID  Recency  Frequency  Monetary  r_score  f_score  m_score  \
0        12346.0      326          1  77183.60        1        1        5   
2        12348.0       76         27   1437.24        2        2        4   
4        12350.0      311         16    294.40        1        2        2   
6        12353.0      205          2     29.30        1        1        1   
8        12355.0      215         11    414.00        1        1        2   
...          ...      ...        ...       ...      ...      ...      ...   
4303     18262.0      141         13    149.48        2        1        1   
4306     18269.0      367          7    168.60        1        1        1   
4313     18278.0       74          9    173.90        2        1        1   
4314     18280.0      278          9    160.70        1        1        1   
4315     18281.0      181          5     46.92        1        1        1   

      rfm_sum      Segment         Country  
0           7  hibernating  United Kingdom  
2           8  hibernating         Finland  
4           5  hibernating          Norway  
6           3  hibernating         Bahrain  
8           4  hibernating         Bahrain  
...       ...          ...             ...  
4303        4  hibernating  United Kingdom  
4306        3  hibernating  United Kingdom  
4313        4  hibernating  United Kingdom  
4314        3  hibernating  United Kingdom  
4315        3  hibernating  United Kingdom  

[1078 rows x 10 columns]

### Customers classified as "Hibernating" due to their long lapse since the last purchase, low spending, and a low number of orders need targeted strategies to re-engage them.

* Launch reactivation campaigns with personalized messages and offers. Highlight new products, changes, or improvements in your offerings to grab their attention.
* Create special comeback offers or discounts exclusively for Hibernating customers.
* Utilize personalized product recommendations based on their historical purchases to showcase items that might interest them. This can reignite their interest in your offerings.